v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,237 @@
+	b.upload.Lock()
+	defer b.upload.Unlock()
+
+	close(b.buffer)
+	oldBuffer := b.buffer
+	b.buffer = make(chan *bufferItem, bufferSizeLimitEvents)
+
+	for event := range oldBuffer {
+		b.push(event)
+	}
+}
+
+// Push attempts to add a new event to the buffer, returning true if an event was dropped.
+// This can be called concurrently.
+func (b *eventBuffer) Push(event *EventV1) {
+	b.push(&bufferItem{EventV1: event})
+}
+
+func (b *eventBuffer) push(event *bufferItem) {
+	// This prevents getting blocked forever writing to a full buffer, in case another routine fills the last space.
+	// Retrying maxEventRetry times to drop the oldest event. Dropping the incoming event if there still isn't room.
+	maxEventRetry := 1000
+
+	for range maxEventRetry {
+		// non-blocking send to the buffer, to prevent blocking if buffer is full so room can be made.
+		select {
+		case b.buffer <- event:
+			return
+		default:
+		}
+
+		// non-blocking drop from the buffer to make room for incoming event.
+		// the buffer could have emptied due to an upload.
+		select {
+		case <-b.buffer:
+			b.incrMetric(logBufferEventDropCounterName)
+		default:
+		}
+	}
+}
+
+// Upload reads events from the buffer and uploads them to the configured client.
+// All the events currently in the buffer are read and written to a gzip compressed JSON array to create a chunk of data.
+// Each chunk is limited by the uploadSizeLimitBytes.
+func (b *eventBuffer) Upload(ctx context.Context) error {
+	b.upload.Lock()
+	defer b.upload.Unlock()
+
+	eventLen := len(b.buffer)
+	if eventLen == 0 {
+		return &bufferEmpty{}
+	}
+
+	encoder := newChunkEncoder(b.uploadSizeLimitBytes)
+
+	for range eventLen {
+		event := b.readEvent()
+		if event == nil {
+			break
+		}
+
+		var result [][]byte
+		if event.chunk != nil {
+			result = [][]byte{event.chunk}
+		} else {
+			serialized, err := b.processEvent(event.EventV1)
+			if err != nil {
+				b.logError("%v", err)
+				continue
+			}
+
+			result, err = encoder.WriteBytes(serialized)
+			if err != nil {
+				b.incrMetric(logEncodingFailureCounterName)
+				b.logError("encoding failure: %v, dropping event with decision ID: %v", err, event.DecisionID)
+				continue
+			}
+		}
+
+		if err := b.uploadChunks(ctx, result); err != nil {
+			return err
+		}
+	}
+
+	// flush any chunks that didn't hit the upload limit
+	result, err := encoder.Flush()
+	if err != nil {
+		b.incrMetric(logEncodingFailureCounterName)
+		b.logError("encoding failure: %v", err)
+		return nil
+	}
+
+	if err := b.uploadChunks(ctx, result); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// uploadChunks attempts to upload multiple chunks to the configured client.
+// In case of failure all the events are added back to the buffer.
+func (b *eventBuffer) uploadChunks(ctx context.Context, result [][]byte) error {
+	var err error
+	for _, chunk := range result {
+		// only attempt an upload if there hasn't been a failure
+		if err == nil {
```

Comment:
If we for some reason end up with a bad chunk on the buffer, that's always rejected by the log server, this runs the risk of backing up the buffer; never letting the consumer-side past the bad chunk. If we instead only pushed the offending chunk to the buffer, and tried to upload the rest, it might be temporarily mistreating a congested log server, but we'd also not make it impossible to upload subsequent good chunks.
As a future improvement, we could attempt identifying if a rejected chunk was bad, or just rejected because of congestion/timeout/etc.

Reply:
Changed so only a failed chunk is pushed back and each chunk gets a fair chance, retry mechanism is a good idea for a future improvement.